bisection.py

```diff
@@ -0,0 +1,373 @@
+                                bmfilter = bmfilter,
+                                workdir = self.workdir)
+        self.output_json = output_json
+
+    # Left: older commit; right: newer commit
+    def regression(self, left: Commit, right: Commit) -> bool:
+        assert left.score is not None
+        assert right.score is not None
+        return left.score - right.score >= self.threshold
+
+    def prep(self) -> bool:
+        if not self.torch_src.prep():
+            return False
+        if not self.torch_src.init_commits(self.start, self.end):
+            return False
+        if not self.bench.prep():
+            return False
+        left_commit = self.torch_src.commits[0]
+        right_commit = self.torch_src.commits[-1]
+        self.bisectq.append((left_commit, right_commit))
+        return True
+        
+    def run(self):
+        while not len(self.bisectq) == 0:
+            (left, right) = self.bisectq.pop(0)
+            left.score = self.bench.get_score(left)
+            right.score = self.bench.get_score(right)
+            if self.regression(left, right):
+                mid = self.torch_src.get_mid_commit(left, right)
+                if mid == None:
+                    self.result.append((left, right))
+                else:
+                    mid.score = self.bench.get_score(mid_commit)
+                    if self.regression(left, mid):
+                        self.bisectq.append(left, mid)
+                    if self.regression(mid, right):
+                        self.bisectq.append(right, mid)
+ 
+    def output(self):
+        json_obj = dict()
+        json_obj["start"] = self.start
+        json_obj["end"] = self.end
+        json_obj["threshold"] = self.threshold
+        json_obj["timeout"] = self.bench.timelimit
+        json_obj["torchbench_branch"] = self.bench.branch
+        json_obj["result"] = []
+        for res in self.result:
+            r = dict()
+            r["commit1"] = res[0].sha
+            r["commit1_time"] = res[0].ctime
+            r["commit1_score"] = res[0].score
+            r["commit2"] = res[1].sha
+            r["commit2_time"] = res[1].ctime
+            r["commit2_score"] = res[1].score
+            json_obj["result"].append(r)
+        with open(self.output_json, 'w') as outfile:
+            json.dump(json_obj, outfile, indent=2)
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser(description=__doc__)
+    parser.add_argument("--work-dir",
+                        help="bisection working directory",
+                        type=exist_dir_path,
+                        required=True)
+    parser.add_argument("--pytorch-src",
+                        help="the directory of pytorch source code git repository",
+                        type=exist_dir_path,
+                        required=True)
+    parser.add_argument("--torchbench-src",
+                        help="the directory of torchbench source code git repository",
```

Comment:
good to mention for each repo whether it will be modified by the script;
and whether it's required to be in a clean state before running.

relatedly, is it very useful to provide your own pytorch repo, or should bisector just clone its own repo into either a tmpdir or just a ~/.torchbench folder it controls?  (at least by default) 

i'm fine with leaving as-is, maybe mark this as a change for future if anything.

Reply:
I think the bisector can clone its own repo, but it can use an option to specify build artifact caches instead (can be ~/.torchbench by default).